quic/s2n-quic-transport/src/space/datagram.rs

```diff
@@ -38,48 +40,101 @@ impl<Config: endpoint::Config> Manager<Config> {
             context,
-            pending_streams: stream_manager.pending_streams(),
+            has_pending_streams: stream_manager.has_pending_streams(),
         };
         self.sender.on_transmit(&mut packet);
     }
 }
+
+impl<Config: endpoint::Config> interest::Provider for Manager<Config> {
+    #[inline]
+    fn transmission_interest<Q: interest::Query>(&self, query: &mut Q) -> interest::Result {
+        if self.sender.has_transmission_interest() {
+            query.on_new_data()?;
+        }
+        Ok(())
+    }
+}
+
 struct Packet<'a, C: WriteContext> {
     context: &'a mut C,
-    pending_streams: bool,
+    has_pending_streams: bool,
 }
 
 const FRAME_TYPE_LEN: usize = 1;
+const MAX_LEN_VALUE: usize = 8;
 
 impl<'a, C: WriteContext> s2n_quic_core::datagram::Packet for Packet<'a, C> {
     /// Returns the remaining space in the packet
     fn remaining_capacity(&self) -> usize {
         let space = self.context.remaining_capacity();
-        //= https://www.rfc-editor.org/rfc/rfc9221#section-4
-        //# The least significant bit of the Type field in the DATAGRAM frame is
-        //# the LEN bit (0x01), which indicates whether there is a Length field
-        //# present: if this bit is set to 0, the Length field is absent and the
-        //# Datagram Data field extends to the end of the packet; if this bit is
-        //# set to 1, the Length field is present.
-        // In the case where the user writes the largest datagram possible
-        // we don't factor in the size of the Length field as it will be the last
-        // frame in the packet. Thus we only have to factor out the frame type field.
-        space - FRAME_TYPE_LEN
+        // Remove the frame type length and the maximum length value
+        space
+            .saturating_sub(FRAME_TYPE_LEN)
+            .saturating_sub(MAX_LEN_VALUE)
     }
 
     /// Writes a single datagram to a packet
-    fn write_datagram(&mut self, data: &[u8]) {
+    fn write_datagram(&mut self, data: &[u8]) -> Result<(), WriteError> {
         let remaining_capacity = self.context.remaining_capacity();
         let data_len = data.len();
         let is_last_frame = remaining_capacity == FRAME_TYPE_LEN + data_len;
+
+        if !is_last_frame {
+            let encoded_length = match VarInt::new(data_len as u64) {
+                Ok(encoded_length) => encoded_length,
+                Err(_e) => return Err(WriteError::DatagramIsTooLarge),
+            };
+            // Calculates the complete size of the encoded datagram, including the
+            // VarInt size and frame size
+            let encoded_datagram_size = encoded_length.encoding_size() + FRAME_TYPE_LEN + data_len;
+
+            if encoded_datagram_size > remaining_capacity {
+                return Err(WriteError::DatagramIsTooLarge);
+            }
+        }
+
         let frame = frame::Datagram {
             is_last_frame,
             data,
         };
         self.context.write_frame(&frame);
```

Comment:
The `write_frame` ensures the packet has enough capacity

```suggestion
        let frame = frame::Datagram {
            is_last_frame,
            data,
        };
        self.context.write_frame(&frame).ok_or(WriteError::DatagramIsTooLarge)?;
```

Reply:
Ah thanks for this suggestion. Just a heads up, I am going to remove the unit test I wrote for this file since it was checking the logic for ensuring the datagram fits in the buffer(since that logic already lives in write_frame).